# gru 生成模型 

参照官方文档    
https://tensorflow.google.cn/tutorials/text/text_generation

## 1数据处理

In [1]:
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf

import numpy as np
import os
import time

In [2]:
path_to_file = tf.keras.utils.get_file('shakespeare.txt', 'https://storage.googleapis.com/download.tensorflow.org/data/shakespeare.txt')
# 读取并为 py2 compat 解码
text = open(path_to_file, 'rb').read().decode(encoding='utf-8')
text = text[:111539]
# 文本长度是指文本中的字符个数
print ('Length of text: {} characters'.format(len(text)))
text[:30]

1122304/1115394 [==============================] - 1s 1us/step
Length of text: 111539 characters


'First Citizen:\nBefore we proce'

In [3]:
# 文本中的非重复字符
vocab = sorted(set(text))
print ('{} unique characters'.format(len(vocab)))


61 unique characters


In [4]:
# 创建从非重复字符到索引的映射
char2idx = {u:i for i, u in enumerate(vocab)}
idx2char = np.array(vocab)

text_as_int = np.array([char2idx[c] for c in text])
text_as_int.shape

(111539,)

In [6]:
text_as_int_target=text_as_int[1:]
print(text_as_int[:100])
print(text_as_int_target[:100])


[16 43 52 53 54  1 13 43 54 43 60 39 48  8  0 12 39 40 49 52 39  1 57 39
  1 50 52 49 37 39 39 38  1 35 48 59  1 40 55 52 54 42 39 52  5  1 42 39
 35 52  1 47 39  1 53 50 39 35 45  7  0  0 11 46 46  8  0 29 50 39 35 45
  5  1 53 50 39 35 45  7  0  0 16 43 52 53 54  1 13 43 54 43 60 39 48  8
  0 34 49 55]
[43 52 53 54  1 13 43 54 43 60 39 48  8  0 12 39 40 49 52 39  1 57 39  1
 50 52 49 37 39 39 38  1 35 48 59  1 40 55 52 54 42 39 52  5  1 42 39 35
 52  1 47 39  1 53 50 39 35 45  7  0  0 11 46 46  8  0 29 50 39 35 45  5
  1 53 50 39 35 45  7  0  0 16 43 52 53 54  1 13 43 54 43 60 39 48  8  0
 34 49 55  1]


In [8]:
# 设定每个输入句子长度的最大值
seq_length = 100

examples_per_epoch = len(text)//(seq_length)
print("examples_per_epoch",examples_per_epoch)
# 从原文的 长串整数编码切出 111500个 字符编码
train_x=text_as_int[:examples_per_epoch*100]
train_y=text_as_int_target[:examples_per_epoch*100]
# reshape成1115行  每行是一个seq
train_x=train_x.reshape([examples_per_epoch,-1])
train_y=train_y.reshape([examples_per_epoch,-1])

print(train_x.shape)
print(train_y.shape)

examples_per_epoch 1115
(1115, 100)
(1115, 100)


In [9]:
# 批大小
BATCH_SIZE = 64

batchs_per_epoch = examples_per_epoch//(BATCH_SIZE)
print("batchs_per_epoch",batchs_per_epoch)

train_x = train_x[:17*64]
train_y = train_y[:17*64]

print(train_x.shape)
print(train_y.shape)

batchs_per_epoch 17
(1088, 100)
(1088, 100)


In [10]:
print(''.join([idx2char[i] for i in train_x[2]]))
print(''.join([idx2char[i] for i in train_y[2]]))

 know Caius Marcius is chief enemy to the people.

All:
We know't, we know't.

First Citizen:
Let us
know Caius Marcius is chief enemy to the people.

All:
We know't, we know't.

First Citizen:
Let us 


## 2模型构建

In [12]:


# 词集的长度
vocab_size = len(vocab)

# 嵌入的维度
embedding_dim = 256

# RNN 的单元数量
rnn_units = 1024


###  model 可以不指定第二维  就是时间步的那一维     

1 train_x 的shape是 1088\*100 (按batch feed的话是64\*100) 行数是句子数 列是指每行有100个char  这里一个char是一个0~vocab_size的编号 代表它在词表里是第几个   
2 模型第一层走一个embedding把 input的每个句子 的最后一维  就是长度是100的那一维 给多拉一维 拉成 100\*256 (256词向量的维度)  意思是给这100个char 每一char都给一个特定的表示性的 char vec   
3 模型第二层 单向gru 要是return_sequences=True的话 每个时间步都会返回  这样的话 这层的输出也是3维  是batchsize seqlen out_h   rnn_units = 1024  故 out_h共 1024个   
4 模型第三层  一个全连接层 而且没激活函数  是让上一层输出的最后一维映射到 61就是词表维度   它的算loss往外放了 没放这

In [13]:
def build_model(vocab_size, embedding_dim, rnn_units, batch_size):
    model = tf.keras.Sequential([
    tf.keras.layers.Embedding(vocab_size, embedding_dim,
                              batch_input_shape=[batch_size, None]),
    tf.keras.layers.GRU(rnn_units,
                        return_sequences=True,
                        stateful=True,
                        recurrent_initializer='glorot_uniform'),
    tf.keras.layers.Dense(vocab_size)
  ])
    return model


In [14]:
model = build_model(
  vocab_size = len(vocab),
  embedding_dim=embedding_dim,
  rnn_units=rnn_units,
  batch_size=BATCH_SIZE)


In [15]:
example_batch_predictions = model(train_x[:64])
print(example_batch_predictions.shape, "# (batch_size, sequence_length, vocab_size)")


(64, 100, 61) # (batch_size, sequence_length, vocab_size)


In [16]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (64, None, 256)           15616     
_________________________________________________________________
gru (GRU)                    (64, None, 1024)          3938304   
_________________________________________________________________
dense (Dense)                (64, None, 61)            62525     
Total params: 4,016,445
Trainable params: 4,016,445
Non-trainable params: 0
_________________________________________________________________


In [17]:
def loss(labels, logits):
    return tf.keras.losses.sparse_categorical_crossentropy(labels, logits, from_logits=True)

example_batch_loss  = loss(train_x[:64], example_batch_predictions)
print("Prediction shape: ", example_batch_predictions.shape, " # (batch_size, sequence_length, vocab_size)")
print("scalar_loss:      ", example_batch_loss.numpy().mean())


Prediction shape:  (64, 100, 61)  # (batch_size, sequence_length, vocab_size)
scalar_loss:       4.1110215


In [18]:
model.compile(optimizer='adam', loss=loss)



In [19]:
# 检查点保存至的目录
checkpoint_dir = './training_checkpoints/gru_ssby_zg/'

# 检查点的文件名
checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt_{epoch}")

checkpoint_callback=tf.keras.callbacks.ModelCheckpoint(
    filepath=checkpoint_prefix,
    save_weights_only=True)



In [20]:
history = model.fit(train_x,train_y,batch_size=BATCH_SIZE, epochs=10, callbacks=[checkpoint_callback])


Epoch 1/10
17/17 [==============================] - 19s 1s/step - loss: 4.0661
Epoch 2/10
17/17 [==============================] - 20s 1s/step - loss: 3.2527
Epoch 3/10
17/17 [==============================] - 21s 1s/step - loss: 2.8938
Epoch 4/10
17/17 [==============================] - 21s 1s/step - loss: 2.5895
Epoch 5/10
17/17 [==============================] - 20s 1s/step - loss: 2.4084
Epoch 6/10
17/17 [==============================] - 20s 1s/step - loss: 2.3025
Epoch 7/10
17/17 [==============================] - 19s 1s/step - loss: 2.2289
Epoch 8/10
17/17 [==============================] - 19s 1s/step - loss: 2.1684
Epoch 9/10
17/17 [==============================] - 19s 1s/step - loss: 2.1122
Epoch 10/10
17/17 [==============================] - 20s 1s/step - loss: 2.0578


## 3文本生成

In [21]:
tf.train.latest_checkpoint(checkpoint_dir)


'./training_checkpoints/gru_ssby_zg/ckpt_10'

In [22]:
model = build_model(vocab_size, embedding_dim, rnn_units, batch_size=1)

model.load_weights(tf.train.latest_checkpoint(checkpoint_dir))

model.build(tf.TensorShape([1, None]))

model.summary()


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (1, None, 256)            15616     
_________________________________________________________________
gru_1 (GRU)                  (1, None, 1024)           3938304   
_________________________________________________________________
dense_1 (Dense)              (1, None, 61)             62525     
Total params: 4,016,445
Trainable params: 4,016,445
Non-trainable params: 0
_________________________________________________________________


In [23]:
idx2char

array(['\n', ' ', '!', '&', "'", ',', '-', '.', ':', ';', '?', 'A', 'B',
       'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
       'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'Y', 'a', 'b', 'c', 'd',
       'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q',
       'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'], dtype='<U1')

In [24]:
test_char = 'w'
test_char_index = char2idx[test_char]
test_char_index

57

In [25]:
test_char_predict = model(np.array([test_char_index]).reshape([1,1]))
test_char_predict

<tf.Tensor: shape=(1, 1, 61), dtype=float32, numpy=
array([[[ 0.6885438 ,  1.3047065 ,  0.369819  , -1.7121626 ,
          0.12830773,  0.8763052 , -0.24802834,  0.50558436,
          0.9085232 ,  0.32237583,  0.31436208, -0.6172425 ,
         -1.6159298 , -0.7094851 , -1.2430074 , -0.23813868,
         -1.582292  , -1.4216698 , -1.5951889 ,  0.5824385 ,
         -1.7450013 , -1.8304125 , -2.3313754 , -1.5038781 ,
         -1.4869186 , -0.9187767 , -1.6996738 , -1.6989669 ,
         -1.5344822 , -0.56702954, -0.94898367,  0.6517089 ,
         -1.4149535 , -1.544923  , -1.6954216 ,  1.4023223 ,
         -0.95612013, -0.29035294, -1.3731216 ,  1.729307  ,
         -0.77102244, -1.1788002 ,  2.5131927 ,  2.0422344 ,
         -1.4027553 , -0.88572955, -0.00840909, -0.45202795,
          0.38620576,  1.0615803 , -0.4932338 , -1.1399279 ,
          0.89263225,  0.40117967,  0.0812659 ,  0.17571734,
         -1.2217294 , -0.2011681 , -0.9924155 ,  0.2572446 ,
         -1.6480094 ]]], dtype=fl

In [26]:
pre_char_index = tf.argmax(test_char_predict.numpy().flatten())
print(pre_char_index,idx2char[pre_char_index])

tf.Tensor(42, shape=(), dtype=int64) h


In [27]:
def generate_text(model, start_string):
  # 评估步骤（用学习过的模型生成文本）

  # 要生成的字符个数
    num_generate = 1000

  # 将起始字符串转换为数字（向量化）
    input_eval = [char2idx[s] for s in start_string]
#     print(input_eval)
    input_eval = tf.expand_dims(input_eval, 0)
    print("input_eval",input_eval)

  # 空字符串用于存储结果
    text_generated = []

  # 低温度会生成更可预测的文本
  # 较高温度会生成更令人惊讶的文本
  # 可以通过试验以找到最好的设定
    temperature = 1.0

  # 这里批大小为 1
    model.reset_states()
    for i in range(num_generate):
        predictions = model(input_eval)
#         print(predictions.shape)
      # 删除批次的维度
        predictions = tf.squeeze(predictions, 0)
#         print(predictions.shape)
        
      # 用分类分布预测模型返回的字符
        predictions = predictions / temperature
        predicted_id=tf.argmax(predictions.numpy().flatten()).numpy()
        print("max predicted_id",predicted_id)
        predicted_id = tf.random.categorical(predictions, num_samples=1)[-1,0].numpy()
        print("random predicted_id",predicted_id)

      # 把预测字符和前面的隐藏状态一起传递给模型作为下一个输入
        input_eval = tf.expand_dims([predicted_id], 0)
        print(input_eval.shape)

        text_generated.append(idx2char[predicted_id])

    return (start_string + ''.join(text_generated))


In [28]:
print(generate_text(model, start_string=u"w"))


input_eval tf.Tensor([[57]], shape=(1, 1), dtype=int32)
max predicted_id 42
random predicted_id 51
(1, 1)
max predicted_id 55
random predicted_id 55
(1, 1)
max predicted_id 52
random predicted_id 46
(1, 1)
max predicted_id 46
random predicted_id 35
(1, 1)
max predicted_id 48
random predicted_id 5
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 41
(1, 1)
max predicted_id 49
random predicted_id 55
(1, 1)
max predicted_id 52
random predicted_id 46
(1, 1)
max predicted_id 46
random predicted_id 46
(1, 1)
max predicted_id 1
random predicted_id 5
(1, 1)
max predicted_id 1
random predicted_id 0
(1, 1)
max predicted_id 30
random predicted_id 34
(1, 1)
max predicted_id 49
random predicted_id 1
(1, 1)
max predicted_id 55
random predicted_id 42
(1, 1)
max predicted_id 35
random predicted_id 35
(1, 1)
max predicted_id 54
random predicted_id 47
(1, 1)
max predicted_id 39
random predicted_id 8
(1, 1)
max predicted_id 1
random predicted_id 0
(1, 1)
max p

max predicted_id 19
random predicted_id 34
(1, 1)
max predicted_id 49
random predicted_id 49
(1, 1)
max predicted_id 55
random predicted_id 55
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 42
random predicted_id 42
(1, 1)
max predicted_id 35
random predicted_id 43
(1, 1)
max predicted_id 53
random predicted_id 37
(1, 1)
max predicted_id 42
random predicted_id 46
(1, 1)
max predicted_id 1
random predicted_id 7
(1, 1)
max predicted_id 0
random predicted_id 1
(1, 1)
max predicted_id 19
random predicted_id 53
(1, 1)
max predicted_id 42
random predicted_id 42
(1, 1)
max predicted_id 39
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 47
(1, 1)
max predicted_id 1
random predicted_id 54
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 46
(1, 1)
max predicted_id 39
random predicted_id 49
(1, 1)
max predicted_id 1
random predicted_id 37
(1, 1)
max predicted_id 39
random predicted_id 42
(1, 1)
max predict

max predicted_id 1
random predicted_id 52
(1, 1)
max predicted_id 1
random predicted_id 36
(1, 1)
max predicted_id 39
random predicted_id 46
(1, 1)
max predicted_id 39
random predicted_id 7
(1, 1)
max predicted_id 0
random predicted_id 0
(1, 1)
max predicted_id 0
random predicted_id 0
(1, 1)
max predicted_id 12
random predicted_id 23
(1, 1)
max predicted_id 15
random predicted_id 28
(1, 1)
max predicted_id 19
random predicted_id 19
(1, 1)
max predicted_id 22
random predicted_id 22
(1, 1)
max predicted_id 19
random predicted_id 19
(1, 1)
max predicted_id 24
random predicted_id 24
(1, 1)
max predicted_id 19
random predicted_id 31
(1, 1)
max predicted_id 29
random predicted_id 29
(1, 1)
max predicted_id 8
random predicted_id 8
(1, 1)
max predicted_id 0
random predicted_id 0
(1, 1)
max predicted_id 19
random predicted_id 19
(1, 1)
max predicted_id 1
random predicted_id 43
(1, 1)
max predicted_id 46
random predicted_id 38
(1, 1)
max predicted_id 1
random predicted_id 53
(1, 1)
max predicted

max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 41
(1, 1)
max predicted_id 49
random predicted_id 42
(1, 1)
max predicted_id 39
random predicted_id 35
(1, 1)
max predicted_id 54
random predicted_id 46
(1, 1)
max predicted_id 46
random predicted_id 38
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 54
(1, 1)
max predicted_id 42
random predicted_id 42
(1, 1)
max predicted_id 39
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 52
(1, 1)
max predicted_id 1
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 49
(1, 1)
max predicted_id 40
random predicted_id 40
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 49
(1, 1)
max predicted_id 40
random predicted_id 55
(1, 1)
max predicted_id 52
random predicted_id 52
(1, 1)
max predicted_id 1
random predicted_id 54
(1, 1)
max predict

max predicted_id 39
random predicted_id 35
(1, 1)
max predicted_id 54
random predicted_id 54
(1, 1)
max predicted_id 1
random predicted_id 54
(1, 1)
max predicted_id 1
random predicted_id 6
(1, 1)
max predicted_id 6
random predicted_id 0
(1, 1)
max predicted_id 12
random predicted_id 30
(1, 1)
max predicted_id 42
random predicted_id 42
(1, 1)
max predicted_id 39
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 38
(1, 1)
max predicted_id 1
random predicted_id 52
(1, 1)
max predicted_id 1
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 46
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 54
(1, 1)
max predicted_id 42
random predicted_id 42
(1, 1)
max predicted_id 39
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 48
(1, 1)
max predicted_id 1
random predicted_id 54
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 59
(1, 1)
max predicted_i

max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 19
(1, 1)
max predicted_id 1
random predicted_id 31
(1, 1)
max predicted_id 29
random predicted_id 29
(1, 1)
max predicted_id 8
random predicted_id 8
(1, 1)
max predicted_id 0
random predicted_id 0
(1, 1)
max predicted_id 33
random predicted_id 12
(1, 1)
max predicted_id 39
random predicted_id 39
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 37
(1, 1)
max predicted_id 49
random predicted_id 42
(1, 1)
max predicted_id 35
random predicted_id 39
(1, 1)
max predicted_id 48
random predicted_id 53
(1, 1)
max predicted_id 1
random predicted_id 2
(1, 1)
max predicted_id 1
random predicted_id 0
(1, 1)
max predicted_id 0
random predicted_id 48
(1, 1)
max predicted_id 49
random predicted_id 49
(1, 1)
max predicted_id 57
random predicted_id 57
(1, 1)
max predicted_id 1
random predicted_id 1
(1, 1)
max predicted_id 54
random predicted_id 57
(1, 1)
max predicted_id